### DISCUSSION TITLE
Trying to detect if there is power voltage with current sensor ACS712 - Raspberry pico W

### DISCUSSION THREAD

**Discussion by Tomply87** [micropython/micropython#11922] [Hardware & Peripherals] [answered]
Hi,
I'm powering a raspberry pico w with a power bank.

The power bank will be always attached to a power supply az-delivery ac-05-3 (220v to 5v)

My goal is to check if 220v is present.

I'm not interested to calculate watt or ampere. 

I can't find a schema or a logic when I try to read from the current sensor

```
from machine import ADC
analog_input_pin = ADC(28)

for x in range(1000):

    ADC.read_u16(analog_input_pin)
```

![Untitled Sketch 2_bb](https://github.com/micropython/micropython/assets/50880593/a348def7-79a3-4417-b74c-76cc85ce91ae)

**Comment by davefes** [comment]
> My goal is to check if 220v is present.

Do you mean you want to measure how much current the AC-05-03 is drawing?

Using a 30Amp sensor to detect 40mA is probably not a good choice.
https://www.seeedstudio.com/blog/2020/02/15/acs712-current-sensor-features-how-it-works-arduino-guide/

**Comment by robert-hh** [comment, marked as answer]
The ACS712 cannot detect voltage, just current. And the quiescent output voltage is Vcc/2, which will be modulated by the pattern of the current, like with +/- x mV, 50Hz. With a few mA on a 30A range sensor, you will always read Vcc/2 and some noise.
I have my doubts whether you can connect the power bank and the Power supply in parallel. Better use two diodes like 1N5817 to decouple them. And then you can feed the output of the power supply through a voltage divider directly into a GPIO pin, which can be checked for low or high. Low means: no power from the power supply, high means power supply on.

**Reply by Tomply87** [reply]
And then you can feed the output of the power supply through a voltage divider directly into a GPIO pin, which can be checked for low or high. Low means: no power from the power supply, high means power supply on.

Can you show me how to do the connection ?

**Comment by bixb922** [comment]
This circuit should apply to the RP2040 too: https://www.upesy.com/blogs/tutorials/measure-voltage-on-esp32-with-adc-with-arduino-code

**Reply by Tomply87** [reply]
Thank you so much!


With this connection

![image](https://github.com/micropython/micropython/assets/50880593/dcfcf29e-c759-4ef9-88e2-11e0e3b00575)

and this code 

MicroPython v1.20.0 on 2023-04-26; Raspberry Pi Pico W with RP2040

It will be necessary to make a cross product to convert the binary value into volts:

Volts = (Volt input / 65535) * binary add value    # 65535 is the maximum value of an unsigned 16-bit integer


```

from machine import Pin, ADC
import time
adc = ADC(Pin(26, mode=Pin.IN))

for x in range(1000):

	value = (adc.read_u16())
	volt = (3.3/65535) * value
	time.sleep(1)
	print(f'{volt} volt')

```





[https://www.upesy.com/blogs/tutorials/micropython-raspberry-pi-pico-adc-usage-measure-voltage](url)

**Comment by robert-hh** [comment]
To tell whether the 5V supply is present or not, you do not need the ADC input. You can use any digital GPIO pin and test for 0 or 1. 

> Can you show me how to do the connection ?

I thought I had added that already.
![power_monitor](https://github.com/micropython/micropython/assets/12476868/16bd8702-c9b8-4533-b9a1-a8f18c3e6ee0)

**Reply by rkompass** [reply]
This is quite similar to the powerchain schematics and explanation in the [Pico-Datasheet](https://datasheets.raspberrypi.com/pico/pico-datasheet.pdf), page 18. The 3.3 above of course are meant to be 3.3K. In the Datasheet they use a voltage divider with 5.6K and 10K , which consumes slightly less current, which in the end makes almost no difference.
I learnt just now that the Pico may be fed with as little as 1.8 Volts, which are then converted to 3.3 Volts.

**Comment by robert-hh** [comment]
Thanks for the hint. It should indeed be 3.3kOhm. I selected a relatively low impedance to cater for the higher leakage current of a Schottky diode compared to a bipolar diode like a 1N400x. But the Schottky diode has a lower forward voltage, which is better here. A quick test with a 1N5817 and 5V showed a leakage current to 2.5µA, the data sheet specified a max. of 500µA at 20V.
Although, given the buck-boost converter of the Pico, a 1N400x would work as well.

**Reply by rkompass** [reply]
Robert indicates it in his drawing, but I just see we should state this also clearly to @Tomply87: You must **not** connect the LiPo Battery directly to a 5V Power supply like in your first drawing. Use a special charger like [this](https://www.sparkfun.com/products/15217) for charging, otherwise your battery might be destroyed very soon.
OTOH:

> These batteries require special charging. Do not attempt to charge these with anything but a specialized Lithium Polymer charger.
Battery includes built-in protection against over voltage, over current, and minimum voltage.

So perhaps it is so well protected that this direct connection is possible again?

Does someone have experience in using such LiPo batteries and reading the available capacity / energy storage? Perhaps with a ADC from the terminal voltage during loading / consuming?

Thank you for the leakage current hint. I once noted that but obviously had forgotten about it.
